Adding warnings for async Tensor serialization in remote and rpc_async

mrshenli opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #34931 Support using self as the destination in rpc.remote for builtin operators
* #34921 Fix dist autograd context Example block format
* #34919 Fix example block format in Distributed Optimizer API doc
* #34914 Fix example format in Distributed Autograd doc
* #34890 Minor fixes for RPC API docs
* #34888 Update descriptions for transmitting CUDA tensors
* #34887 Removing experimental tag in for RPC and adding experimental tag for RPC+TorchScript
* **#34885 Adding warnings for async Tensor serialization in remote and rpc_async**
* #34884 Add a warning for RRef serialization

Differential Revision: [D20491279](https://our.internmc.facebook.com/intern/diff/D20491279)

mrshenli committed:
Adding warnings for async Tensor serialization in remote and rpc_async

[ghstack-poisoned]

mrshenli committed:
Update on "Adding warnings for async Tensor serialization in remote and rpc_async"




[ghstack-poisoned]

pritamdamania87 commented on torch/distributed/rpc/api.py:
```diff
@@ -370,6 +370,13 @@ def remote(to, func, args=None, kwargs=None):
+        sending them over the wire, which could be done by a different thread
+        depending on the RPC backend type. The caller should make sure that the
+        contents of those tensors stay intact until the returned RRef is
+        confirmed by the owner.
```
The user has no way currently to know when the RRef has been confirmed by the owner right? Should we mention that either the user should never modify the passed in Tensors or only modify them after calling RRef.to_here()? This does add some restrictions to the API, but I'm not sure if there is a better way.

mrshenli replied:
[`RRef.confirmed_by_owner()`](https://pytorch.org/docs/master/rpc.html#torch.distributed.rpc.RRef.confirmed_by_owner) is a public API. Let me add a link to that API here.

pritamdamania87 replied:
Aaah didn't know about that! Sounds good to me!

mrshenli committed:
Update on "Adding warnings for async Tensor serialization in remote and rpc_async"


Differential Revision: [D20491279](https://our.internmc.facebook.com/intern/diff/D20491279)

[ghstack-poisoned]

mrshenli committed:
Update on "Adding warnings for async Tensor serialization in remote and rpc_async"


Differential Revision: [D20491279](https://our.internmc.facebook.com/intern/diff/D20491279)

[ghstack-poisoned]
